docs/RFCS/online_schema_change.md

```diff
@@ -0,0 +1,166 @@
+
+```go
+  startKey := MakeTableIndexKey(desc.ID, desc.Primary)
+  endKey := startKey.PrefixEnd()
+  for startKey != endKey {
+    var lastKey roachpb.Key
+    err := db.Txn(func(txn *Txn) error {
+      txn.SetPriority(VeryLowPriority)
+      scan, err := txn.Scan(startKey, endKey, 1000)
+      if err != nil {
+        return err
+      }
+      lastKey = getLastKeyOfFullRowInScan(scan)
+      b := makeIndexKeysBatch(scan)
+      return txn.CommitInBatch(b)
+    })
+    if err != nil {
+      // Abort!
+    }
+    startKey = lastKey.Next()
+  }
+```
+
+The above pseudo-code is intended to give the gist of how backfilling
+will work. We'll iterate over the primary index transactionally
+retrieving keys and generating index entries. We perform the
+backfilling work transactionally in order to avoid anomalies involving
+deletions. Backfilling might duplicate work performed by concurrent
+insertions or updates, but the blind overwrites of identical data are
+safe.
+
+Since schema change operations are potentially long running they need
+to be restartable or abortable if the node performing them dies. We
+accomplish this by performing the schema change operation for a table
+on a well known node: the replica holding the leader lease for the
+first range of the table (i.e. containing the key `/<tableID>`). When
+a node receives a schema change operation such as `CREATE INDEX` it
+will forward the operation to this "table leader". When the table
+leader restarts it will load the associated table descriptor and
+restart or abort the schema change operation. Note that aborting a
+schema change operation needs to maintain the invariant that the
+descriptor version only increase.
+
+For initial implementation simplicity, we will only allow a single
+schema change operation at a time per table. This restriction can be
+lifted in the future if we see a benefit in allowing concurrent schema
+changes on a table (e.g. concurrently adding multiple indexes).
+
+# Drawbacks
+
+* None. This is rock solid.
+
+# Alternatives
+
+* No real good ones. We could try to come up with some sort of global
+  table lock mechanism and then synchronously perform the
+  backfill. This seems as difficult as the current proposal and much
+  worse for the user experience.
+
+# Unresolved questions
+
+* Is there a way to avoid performing the backfilling work
+  transactionally? Is it worth optimizing? Delete, in particular,
+  seems problematic. If we scan a portion of the primary index
+  non-transactionally and then generate index keys, a concurrent
+  delete can come in and delete one of the rows and not generate a
+  delete of the index key unless we left a "tombstone" deletion for
+  the index key.
+
+* If the node performing the backfill gets restarted we should figure
+  out a way to avoid restarting the backfill from scratch. One thought
+  is that the backfill operation can periodically checkpoint the high
+  water mark of its progress (perhaps in the descriptor itself).
```

Comment:
you do talk about checkpointing here. We could add it to the descriptor but we are not allowed to change the version. I'd add it as another column to the descriptor table